### PR TITLE
CP-20603: Pass vGPU fd to xenguest on source host during migration

### PR DESCRIPTION
**Pull request by robhoes**
...and additional updates.

### CODE REVIEW

File: lib/xenops_server.ml [hunk 1]
```diff
@@ -96,7 +96,7 @@ type atomic =
 	| VM_shutdown_domain of (Vm.id * shutdown_request * float)
 	| VM_s3suspend of Vm.id
 	| VM_s3resume of Vm.id
-	| VM_save of (Vm.id * flag list * data)
+	| VM_save of (Vm.id * flag list * data * data option)
```
Comment (thomassa) [inline lib/xenops_server.ml]: Is it worth a comment describing the uses/meanings of the elements in this four-tuple anonymous type?
(Using a record with named fields is probably over the top.)

File: xc/device.ml [hunk 2]
```diff
@@ -1746,7 +1746,8 @@ let start_vgpu ~xs task ?restore_fd domid vgpus vcpus =
 					[uuid, fd]
 			in
 			let args = vgpu_args_of_nvidia domid vcpus vgpu fds in
-			let ready_path = Printf.sprintf "/local/domain/%d/vgpu-pid" domid in
+			let ready_path = Printf.sprintf "/local/domain/%d/vgpu/state" domid in
+			let ready_val = if restore_fd = None then "running" else "restoring" in
```
Comment (thomassa) [inline xc/device.ml]: What is going to read this path? Can we define the path and the valid values in a single place, shared by this code and the reader?

Comment (robhoes) [inline xc/device.ml]: Oops, I've made a rebase mistake: this bit goes away in the next patch.

File: xc/domain.ml [hunk 3]
```diff
@@ -1066,9 +1066,18 @@ let restore_vgpu (task: Xenops_task.t) ~xc ~xs domid fd vgpu vcpus =
 
 type suspend_flag = Live | Debug
 
-let write_libxc_record (task: Xenops_task.t) ~xc ~xs ~hvm xenguest_path domid uuid fd flags progress_callback qemu_domid do_suspend_callback =
+let write_libxc_record (task: Xenops_task.t) ~xc ~xs ~hvm xenguest_path domid uuid fd vgpu_fd flags progress_callback qemu_domid do_suspend_callback =
```
Comment (thomassa) [inline xc/domain.ml]: I'm slightly uncomfortable with having a string of nine unlabelled parameters to be supplied in the right order, especially with two consecutive params having the same type (`fd` and `vgpu_fd`).

File: xc/device.ml [hunk 4]
```diff
@@ -1746,15 +1749,25 @@ let start_vgpu ~xs task ?restore_fd domid vgpus vcpus =
...
 			let vgpu_pid = init_daemon ~task ~path:!Xc_resources.vgpu ~args
-				~name:"vgpu" ~domid ~xs ~ready_path ~timeout:!Xenopsd.vgpu_ready_timeout
+				~name:"vgpu" ~domid ~xs ~ready_path:state_path ~timeout:!Xenopsd.vgpu_ready_timeout
 				~cancel ~fds () in
 			Forkhelpers.dontwaitpid vgpu_pid
 		end else
-			D.info "Daemon %s is already running for domain %d" !Xc_resources.vgpu domid
+			info "Daemon %s is already running for domain %d" !Xc_resources.vgpu domid;
+
+		(* Keep waiting until DEMU's state becomes "running", which means that is it
```
Comment (thomassa) [inline xc/device.ml]: Swap "is" and "it" in comment.

File: lib/xenops_server.ml [hunk 5]
```diff
@@ -2186,8 +2181,7 @@ module VM = struct
 					end else
 						Some (queue_operation dbg vm_id op)
 				in
-				Opt.iter (fun t -> t |> Xenops_client.wait_for_task dbg |> ignore) task;
-				Xenops_migrate.(Handshake.send transferred_fd Handshake.Success)
```
Comment (thomassa) [inline lib/xenops_server.ml]: Why remove this handshake? Has it moved to somewhere else?

Comment (robhoes) [inline lib/xenops_server.ml]: It's not needed anymore due to the parallel construction of `VM_save` and `VM_save_vgpu`, which synchronises both operations. So the handshake at "Synchronisation point 2" is all we need now.

**Comment by robhoes** [comment]
@thomassa Thanks, I've updated it.

**Pull request merged**